chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, ?height, "can't find block hash by height");
+                return FlatStorageReshardingTaskStatus::Failed;
+            }
+        }
+
+        // Apply deltas and then create the flat storage.
+        let apply_result =
+            self.shard_catchup_apply_deltas(shard_uid, flat_head_block_hash, chain_store);
+        let Ok((num_batches_done, flat_head)) = apply_result else {
+            error!(target: "resharding", ?shard_uid, err = ?apply_result.unwrap_err(), "flat storage shard catchup delta application failed!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        match self.shard_catchup_finalize_storage(shard_uid, &flat_head) {
+            Ok(_) => {
+                let task_status = FlatStorageReshardingTaskStatus::Successful { num_batches_done };
+                info!(target: "resharding", ?shard_uid, ?task_status, "flat storage shard catchup task finished");
+                // At this point we can trigger the reload of memtries.
+                self.scheduler.send(ReshardingRequest::MemtrieReload { shard_uid });
+                task_status
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, "flat storage shard catchup finalize failed!");
+                FlatStorageReshardingTaskStatus::Failed
+            }
+        }
+    }
+
+    /// Applies flat storage deltas in batches on a shard that is in catchup status.
+    ///
+    /// Returns the number of delta batches applied and the final tip of the flat storage.
+    fn shard_catchup_apply_deltas(
+        &self,
+        shard_uid: ShardUId,
+        mut flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> Result<(usize, Tip), Error> {
+        const CATCH_UP_BLOCKS: u32 = 50;
+
+        let mut num_batches_done: usize = 0;
+
+        loop {
+            let _span = tracing::debug_span!(
+                target: "resharding",
+                "shard_catchup_apply_deltas/batch",
+                ?shard_uid,
+                ?flat_head_block_hash,
+                batch_id = ?num_batches_done)
+            .entered();
+            let chain_final_head = chain_store.final_head()?;
+            let mut merged_changes = FlatStateChanges::default();
+            let store = self.runtime.store().flat_store();
+            let mut store_update = store.store_update();
+
+            // Merge deltas from the next blocks until we reach chain final head.
+            for _ in 0..CATCH_UP_BLOCKS {
+                let height = chain_store.get_block_height(&flat_head_block_hash)?;
+                if height > chain_final_head.height {
+                    let msg = format!("flat head moved too far: head = {flat_head_block_hash}, height = {height}, final block height = {}", chain_final_head.height);
+                    return Err(Error::ReshardingError(msg));
+                }
```

Comment:
Can this happen or is it just in case?

Reply:
I'm afraid this can happen in the presence of forks. To be confirmed with a resharding-with-forks test..

Reply:
Stated in comment [here](https://github.com/near/nearcore/pull/12312#discussion_r1820701423), I guess we can avoid forks processing etc. altogether if we start our flat storage split AFTER split height <= chain head

Reply:
Hmm I'm not sure we can start flat storage split AFTER split height <= chain head. I mean, we can definitely do that but then the resulting flat head will be beyond chain head because of deltas.
The only way I see to keep flat head <= chain head is to split deltas into deltas instead of split deltas into flat state directly. 
I know it sounds confusing, we can take it offline

Reply:
On second thought, you were right, this path can no longer happen